src/identity/identity.ts

```diff
@@ -0,0 +1,355 @@
+
+    // prevent wrong sharing types
+    if (type !== 'read' && type !== 'readWrite') {
+      throw new Error(`Unknown access type passed to "identity.grantAccess(${identity}, ${type}, ${note})"`);
+    }
+
+    // load addressbook to check for changes
+    const addressbook = await profile.getAddressBook();
+    const originContact = { ...addressbook.profile[identity] };
+
+    // save information into the current profile about the grant status
+    await Promise.all([
+      profile.addProfileKey(identity, 'hasIdentityAccess', type),
+      profile.addProfileKey(identity, 'identityAccessGranted', Date.now().toString()),
+      profile.addProfileKey(identity, 'identityAccessNote', note),
+    ]);
+
+    // stack the promises, so we can Promise.all them at the end
+    const promises = [
+      // store the newly added / updated profile keys
+      profile.storeForAccount(profile.treeLabels.addressBook),
+    ];
+
+    // only send bmail, if the permissions have changed
+    if (!originContact || originContact.hasIdentityAccess !== type) {
+      // send a b-mail including the activeIdentity encryptionKey to the target identity
+      promises.push(mailbox.sendMail(
+        {
+          content: {
+            from: activeIdentity,
+            fromAlias: bmailContent.fromAlias,
+            title: bmailContent.title,
+            body: bmailContent.body,
+            attachments: [
+              {
+                encryptionKey: runtimeConfig.keyConfig[web3.utils.soliditySha3(activeIdentity)],
+                permission: type,
+                type: 'identityAccess',
+              },
+            ],
+          },
+        },
+        activeIdentity,
+        identity,
+      ));
+    }
+
+    // grant access to act on behalf of the identity (only if the permission wasn't added before)
+    if (type === 'readWrite' && originContact.hasIdentityAccess !== type) {
+      promises.push(this.grantWriteAccess(identity));
+    }
+
+    // wait until everything is saved to the chain
+    await Promise.all(promises);
+  }
+
+  /**
+   * Adds a identity to the did (pubKey + authorization) and adds the identity to the current
+   * activeIdentities keyholder.
+   *
+   * @param      {string}  identity  identity to give write access to.
+   */
+  public async grantWriteAccess(identity: string): Promise<void> {
+    await this.ensureOwnerAndIdentityProfile();
+
+    const {
+      activeIdentity,
+      contractLoader,
+      did,
+      executor,
+      nameResolver,
+      underlyingAccount,
+    } = this.options;
+
+    // get the did for the current activeIdentity
+    const activeDidAddress = await did.convertIdentityToDid(activeIdentity);
+    const didDocumentToUpdate = await did.getDidDocument(activeDidAddress);
+    // add the new identity to the did document
+    didDocumentToUpdate.publicKey.push({
+      id: `${didDocumentToUpdate.id}#${identity}`,
+      type: 'Secp256k1VerificationKey2018',
+      controller: didDocumentToUpdate.id,
```

Comment:
controller should be the did of the identity

Reply:
See https://testcore.evan.network/did/did:evan:testcore:0xfb561bf63c29f6561664ceb185a9e3adc80ce2bf